src/vumi2/applications/turn_channels_api/turn_channels_api.py

```diff
@@ -0,0 +1,214 @@
+    "Timed out sending event after %(timeout)s seconds. Event: %(event)s"
+)
+LOG_API_ERR = (
+    "Error sending message, received HTTP code %(code)s with"
+    " error %(error)s. Message: %(message)s"
+)
+
+logger = getLogger(__name__)
+
+
+@define(kw_only=True)
+class TurnChannelsApiConfig(BaseConfig):
+    # AMQP connector name. This is also used as `transport_name` for non-reply
+    # outbound messages and `channel_id` for events.
+    connector_name: str
+
+    # Base URL path for HTTP requests.
+    vumi_base_url_path: str = ""
+
+    # Base URL path for requests to Turn.
+    turn_base_url_path: str = ""
+
+    # Auth token for requests to Turn.
+    auth_token: str
+
+    # This is a required `Message` field, so we need to set it for non-reply
+    # outbound messages.
+    transport_type: TransportType = TransportType.SMS
+
+    # All outbound messages must be replies or have a from address.
+    default_from_addr: str
+
+    # Maximum time allowed (in seconds) for outbound message request handling.
+    request_timeout: float = 4 * 60
+
+    # Maximum time allowed (in seconds) for inbound message and event HTTP
+    # requests.
+    mo_message_url_timeout: float = 10
+    event_url_timeout: float = 10
+
+    # Secret key used to sign outbound messages.
+    secret_key: str
+
+    def vumi_url(self, path: str) -> str:
+        return "/".join([self.vumi_base_url_path.rstrip("/"), path.lstrip("/")])
+
+
+class TurnChannelsApi(BaseWorker):
+    """
+    An implementation of the Turn Channels API.
+
+    Inbound messages and events are sent over HTTP to the configured
+    URL(s). Outbound messages are received over HTTP and sent to the
+    configured transport.
+    """
+
+    config: TurnChannelsApiConfig
+
+    async def setup(self) -> None:
+        self.connector = await self.setup_receive_inbound_connector(
+            self.config.connector_name,
+            self.handle_inbound_message,
+            self.handle_event,
+        )
+        try:
+            self.http.app.add_url_rule(
+                self.config.vumi_url("/messages"),
+                view_func=self.http_send_message,
+                methods=["POST"],
+            )
+        except Exception as e:
+            logger.exception(e)
+            raise e
+        await self.start_consuming()
+
+    async def handle_inbound_message(self, message: Message) -> None:
+        """
+        Send the vumi message as an HTTP request to the configured URL.
+        """
+        logger.debug("Consuming inbound message %s", message)
+        msg = turn_inbound_from_msg(message, message.transport_name)
+
+        headers = {}
+
+        headers["Authorization"] = f"Bearer {self.config.auth_token}"
+
+        timeout = self.config.mo_message_url_timeout
+        with move_on_after(timeout) as cs:
+            async with AsyncClient() as client:
+                resp = await client.post(
+                    self.config.turn_base_url_path.format(message.message_id),
```

Comment:
I'm not sure this is correct. The URL for inbound messages is `https://whatsapp.turn.io/v1/numbers/<number_id>/messages`. So if we say have `config.turn_base_url_path`, set to `https://whatsapp.turn.io/v1/numbers/<number_id>`, then we'll need to be adding `/messages` to the end of the pass, not adding a message ID into it.